File: leetcode/692. Top K Frequent Words/priority_queue.py
class Solution {
public:
    vector<string> topKFrequent(vector<string>& words, int k) {
        ios::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);
        
        unordered_map<string, int> mp;        
        for (auto word: words) mp[word]++;
        
        struct cmp {
            bool operator()(pair<int, string> &a, pair<int, string> &b) {
                return a.first == b.first ? a.second < b.second : a.first > b.first;
            }
        };
        
        priority_queue<pair<int, string>, vector<pair<int, string>>, cmp> pq;
        for (auto it : mp) {
            pq.push({it.second, it.first});
            if (pq.size() > k) pq.pop();
        }
        
        vector<string> ans(k);
        while(k) {
            ans[--k] = pq.top().second;
            pq.pop();
        }
        return ans;
    }
};

File: leetcode/1962. Remove Stones to Minimize the Total/priority_queue.cpp
class Solution {
public:
    int minStoneSum(vector<int>& piles, int k) {
        priority_queue<int> pq;
        
        int total = 0;
        for (auto it : piles) {
            total += it;
            pq.push(it);
        }
        
        int remove = 0;
        while (k--) {
            int t = pq.top();
            pq.pop();
            int d = t / 2;
            remove += d;
            pq.push(t - d);
        }
        
        return total - remove;
    }
};